compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -363,6 +363,35 @@ object desugar {
     adaptToExpectedTpt(tree)
   }
 
+  /** Split out the quoted pattern type variable definition from the pattern.
+   *
+   *  Type variable definitions are all the `type t` defined at the start of a quoted pattern.
+   *  Were name `t` is a pattern type variable name (i.e. lower case letters).
+   *
+   *  ```
+   *   type t1; ...; type tn; <pattern>
+   *  ```
+   *  is split into
+   *  ```
+   *   (List(<type t1>; ...; <type tn>), <pattern>)
+   *  ```
+   */
+  def quotedPatternTypeVariables(tree: untpd.Tree)(using Context): (List[untpd.TypeDef], untpd.Tree) =
+    tree match
+      case untpd.Block(stats, expr) =>
+        val untpdTypeVariables = stats.takeWhile {
+          case tdef @ untpd.TypeDef(name, _) => name.isVarPattern
+          case _ => false
+        }.asInstanceOf[List[untpd.TypeDef]]
+        val otherStats = stats.dropWhile {
+          case tdef @ untpd.TypeDef(name, _) => name.isVarPattern
+          case _ => false
+        }
```

Comment:
In the case of quoted type patterns, it might be equivalent. But this might not be the case with expression quote patterns. Consider this ill-formed pattern `'{ type t; $x; type u; $y }`, in this case we only want to desugar `type t` as a type variable. When checking the pattern `tpye u` will chase an error. Changing the way we partition would the way we emit errors in those ill-formed patterns.

Reply:
> in this case we only want to desugar type t as a type variable. 

Right, this should be achievable by using `.span` instead of `.partition`.